File: src/gleam_crypto_ecdsa_ffi.mjs
import { BitArray, Ok, Error as ResultError } from "./gleam.mjs";
import {
  Sha1,
  Sha224,
  Sha256,
  Sha384,
  Sha512,
  Md5,
  Pem,
  Der,
} from "./gleam/crypto.mjs";
import { P256, P384, P521 } from "./gleam/crypto/ecdsa.mjs";
import * as crypto from "node:crypto";

function algorithmName(algorithm) {
  if (algorithm instanceof Sha1) {
    return "sha1";
  } else if (algorithm instanceof Sha224) {
    return "sha224";
  } else if (algorithm instanceof Sha256) {
    return "sha256";
  } else if (algorithm instanceof Sha384) {
    return "sha384";
  } else if (algorithm instanceof Sha512) {
    return "sha512";
  } else if (algorithm instanceof Md5) {
    return "md5";
  } else {
    throw new Error("Unsupported algorithm");
  }
}

function formatName(format) {
  if (format instanceof Pem) {
    return "pem";
  } else if (format instanceof Der) {
    return "der";
  } else {
    throw new Error("Unsupported format");
  }
}

function curveName(curve) {
  if (curve instanceof P256) {
    return "prime256v1";
  } else if (curve instanceof P384) {
    return "secp384r1";
  } else if (curve instanceof P521) {
    return "secp521r1";
  } else {
    throw new Error("Unsupported curve");
  }
}

// --- Key Generation ---

export function generateKeyPair(curve) {
  try {
    const { publicKey, privateKey } = crypto.generateKeyPairSync("ec", {
      namedCurve: curveName(curve),
    });
    return new Ok([privateKey, publicKey]);
  } catch (e) {
    return new ResultError(undefined);
  }
}

// --- Key Import ---

export function privateKey(data, format) {
  try {
    const fmt = formatName(format);
    const key = crypto.createPrivateKey({
      key: data.rawBuffer,
      format: fmt,
      type: fmt === "pem" ? undefined : "sec1",
    });
    return new Ok(key);
  } catch (e) {
    // Try PKCS8 format for DER
    if (format instanceof Der) {
      try {
        const key = crypto.createPrivateKey({
          key: data.rawBuffer,
          format: "der",
          type: "pkcs8",
        });
        return new Ok(key);
      } catch (e2) {
        return new ResultError(undefined);
      }
    }
    return new ResultError(undefined);
  }
}

export function publicKey(data, format) {
  try {
    const fmt = formatName(format);
    const key = crypto.createPublicKey({
      key: data.rawBuffer,
      format: fmt,
      type: fmt === "pem" ? undefined : "spki",
    });
    return new Ok(key);
  } catch (e) {
    return new ResultError(undefined);
  }
}

// --- Key Export ---

export function privateKeyToBytes(key, format) {
  const fmt = formatName(format);
  const exported = key.export({
    format: fmt,
    type: "sec1",
  });
  if (fmt === "pem") {
    return new BitArray(new Uint8Array(Buffer.from(exported)));
  }
  return new BitArray(new Uint8Array(exported));
}

export function publicKeyToBytes(key, format) {
  const fmt = formatName(format);
  const exported = key.export({
    format: fmt,
    type: "spki",
  });
  if (fmt === "pem") {
    return new BitArray(new Uint8Array(Buffer.from(exported)));
  }
  return new BitArray(new Uint8Array(exported));
}

// --- Public Key Derivation ---

export function publicKeyFromPrivate(privateKey) {
  return crypto.createPublicKey(privateKey);
}

// --- Sign/Verify ---

export function sign(key, message, hash) {
  const algo = algorithmName(hash);
  const signature = crypto.sign(algo, message.rawBuffer, key);
  return new BitArray(new Uint8Array(signature));
}

export function verify(key, message, signature, hash) {
  const algo = algorithmName(hash);
  try {
    return crypto.verify(algo, message.rawBuffer, key, signature.rawBuffer);
  } catch (e) {
    return false;
  }
}

File: src/gleam_crypto_ecdsa_ffi.erl
-module(gleam_crypto_ecdsa_ffi).

-include_lib("public_key/include/public_key.hrl").

-export([generate_key_pair/1, private_key/2, public_key/2, private_key_to_bytes/2,
         public_key_to_bytes/2, public_key_from_private/1, sign/3, verify/4]).

convert_algorithm(sha1) ->
    sha;
convert_algorithm(Algorithm) ->
    Algorithm.

curve_name(p256) ->
    secp256r1;
curve_name(p384) ->
    secp384r1;
curve_name(p521) ->
    secp521r1.

oid_to_curve(?secp256r1) ->
    secp256r1;
oid_to_curve(?secp384r1) ->
    secp384r1;
oid_to_curve(?secp521r1) ->
    secp521r1.

curve_to_oid(secp256r1) ->
    ?secp256r1;
curve_to_oid(secp384r1) ->
    ?secp384r1;
curve_to_oid(secp521r1) ->
    ?secp521r1.

%% --- Key Generation ---

generate_key_pair(Curve) ->
    try
        CurveName = curve_name(Curve),
        {PubPoint, PrivKey} = crypto:generate_key(ecdh, CurveName),
        {ok, {{ecdsa_private, PrivKey, PubPoint, CurveName}, {ecdsa_public, PubPoint, CurveName}}}
    catch
        _:_ ->
            {error, nil}
    end.

%% --- Key Import ---

private_key(Data, Format) ->
    try
        case Format of
            pem ->
                [Entry | _] = public_key:pem_decode(Data),
                Key = public_key:pem_entry_decode(Entry),
                private_key_from_record(Key);
            der ->
                Key = public_key:der_decode('ECPrivateKey', Data),
                private_key_from_record(Key)
        end
    catch
        _:_ ->
            {error, nil}
    end.

private_key_from_record(#'ECPrivateKey'{privateKey = PrivKey,
                                        parameters = {namedCurve, OID},
                                        publicKey = PubKey}) ->
    PrivKeyBin =
        case is_list(PrivKey) of
            true ->
                list_to_binary(PrivKey);
            false ->
                PrivKey
        end,
    CurveName = oid_to_curve(OID),
    PubPoint =
        case PubKey of
            undefined ->
                {DerivedPub, _} = crypto:generate_key(ecdh, CurveName, PrivKeyBin),
                DerivedPub;
            asn1_NOVALUE ->
                {DerivedPub, _} = crypto:generate_key(ecdh, CurveName, PrivKeyBin),
                DerivedPub;
            _ ->
                PubKey
        end,
    {ok, {ecdsa_private, PrivKeyBin, PubPoint, CurveName}};
private_key_from_record(#'PrivateKeyInfo'{privateKeyAlgorithm =
                                              #'PrivateKeyInfo_privateKeyAlgorithm'{algorithm =
                                                                                        ?'id-ecPublicKey',
                                                                                    parameters =
                                                                                        Params},
                                          privateKey = PrivKeyDer}) ->
    {asn1_OPENTYPE, ParamsDer} = Params,
    {ok, OID} = 'OTP-PUB-KEY':decode('EcpkParameters', ParamsDer),
    Key = public_key:der_decode('ECPrivateKey', PrivKeyDer),
    private_key_from_record(Key#'ECPrivateKey'{parameters = {namedCurve, OID}});
private_key_from_record(_) ->
    {error, nil}.

public_key(Data, Format) ->
    try
        case Format of
            pem ->
                [Entry | _] = public_key:pem_decode(Data),
                Key = public_key:pem_entry_decode(Entry),
                public_key_from_record(Key);
            der ->
                Key = public_key:der_decode('SubjectPublicKeyInfo', Data),
                public_key_from_record(Key)
        end
    catch
        _:_ ->
            {error, nil}
    end.

public_key_from_record(#'SubjectPublicKeyInfo'{algorithm =
                                                   #'AlgorithmIdentifier'{algorithm =
                                                                              ?'id-ecPublicKey',
                                                                          parameters = Params},
                                               subjectPublicKey = PubKey}) ->
    %% Parameters can be DER-encoded bytes or already decoded {namedCurve, OID}
    {namedCurve, OID} =
        case Params of
            {namedCurve, _} ->
                Params;
            _ ->
                public_key:der_decode('EcpkParameters', Params)
        end,
    CurveName = oid_to_curve(OID),
    {ok, {ecdsa_public, PubKey, CurveName}};
%% Format returned by public_key:pem_entry_decode for EC public keys
public_key_from_record({{'ECPoint', PubKey}, {namedCurve, OID}}) ->
    CurveName = oid_to_curve(OID),
    {ok, {ecdsa_public, PubKey, CurveName}};
public_key_from_record({{#'ECPoint'{point = PubKey}, {namedCurve, OID}}, _}) ->
    CurveName = oid_to_curve(OID),
    {ok, {ecdsa_public, PubKey, CurveName}};
public_key_from_record(_) ->
    {error, nil}.

%% --- Key Export ---

private_key_to_bytes({ecdsa_private, PrivKey, PubKey, CurveName}, Format) ->
    OID = curve_to_oid(CurveName),
    Record =
        #'ECPrivateKey'{version = 1,
                        privateKey = binary_to_list(PrivKey),
                        parameters = {namedCurve, OID},
                        publicKey = PubKey},
    case Format of
        pem ->
            PemEntry = public_key:pem_entry_encode('ECPrivateKey', Record),
            public_key:pem_encode([PemEntry]);
        der ->
            public_key:der_encode('ECPrivateKey', Record)
    end.

public_key_to_bytes({ecdsa_public, PubKey, CurveName}, Format) ->
    OID = curve_to_oid(CurveName),
    Record =
        #'SubjectPublicKeyInfo'{algorithm =
                                    #'AlgorithmIdentifier'{algorithm = ?'id-ecPublicKey',
                                                           parameters = {namedCurve, OID}},
                                subjectPublicKey = PubKey},
    case Format of
        pem ->
            PemEntry = public_key:pem_entry_encode('SubjectPublicKeyInfo', Record),
            public_key:pem_encode([PemEntry]);
        der ->
            public_key:der_encode('SubjectPublicKeyInfo', Record)
    end.

%% --- Public Key Derivation ---

public_key_from_private({ecdsa_private, _PrivKey, PubKey, CurveName}) ->
    {ecdsa_public, PubKey, CurveName}.

%% --- Sign/Verify ---

sign({ecdsa_private, PrivKey, _PubKey, CurveName}, Message, Hash) ->
    Algorithm = convert_algorithm(Hash),
    crypto:sign(ecdsa, Algorithm, Message, [PrivKey, CurveName]).

verify({ecdsa_public, PubKey, CurveName}, Message, Signature, Hash) ->
    Algorithm = convert_algorithm(Hash),
    try
        crypto:verify(ecdsa, Algorithm, Message, Signature, [PubKey, CurveName])
    catch
        _:_ -> false
    end.
